examples/iso_dh/DY.Example.DH.SecurityProperties.fst

```diff
@@ -0,0 +1,175 @@
+open DY.Example.DH.Protocol.Stateful.Proof
+
+#set-options "--fuel 8 --ifuel 8 --z3rlimit 25  --z3cliopt 'smt.qi.eager_threshold=100'"
+
+
+(*** Authentication Properties ***)
+
+val initiator_authentication:
+  tr:trace -> i:nat ->
+  alice:principal -> bob:principal ->
+  gx:bytes -> gy:bytes -> k:bytes ->
+  Lemma
+  (requires 
+    trace_invariant tr /\
+    event_triggered_at tr i alice (Initiate2 alice bob gx gy k)
+  )
+  (ensures 
+    (exists alice_si. is_corrupt tr (principal_state_label alice alice_si)) \/ 
+    is_corrupt tr (principal_label bob) \/
+    (exists y. event_triggered (prefix tr i) bob (Respond1 alice bob gx gy y) /\
+    k == dh y gx)
+  )
+let initiator_authentication tr i alice bob gx gy k = ()
+
+val responder_authentication: 
+  tr:trace -> i:nat ->
+  alice:principal -> bob:principal ->
+  gx:bytes -> gy:bytes -> k:bytes ->
+  Lemma
+  (requires 
+    trace_invariant tr /\
+    event_triggered_at tr i bob (Respond2 alice bob gx gy k)
+  )
+  (ensures 
+    is_corrupt tr (principal_label alice) \/ 
+    (exists bob_si. is_corrupt tr (principal_state_label bob bob_si)) \/
+    event_triggered (prefix tr i) alice (Initiate2 alice bob gx gy k)
+  )
+let responder_authentication tr i alice bob gx gy k = ()
+
+(*** Forward Secrecy Properties ***)
+
+/// This lemma is needed to proof the forward secrecy
+/// security properties.
+/// It is never explicitly called but automatically
+/// instantiated via the SMTPat.
+/// In the forward secrecy security property the problem
+/// is that we do not explicitly have the session id
+/// available. That is the reason for using the SMTPat.
+///
+/// Alternatively the SMTPat of the lemma 
+/// principal_flow_to_principal_state in DY.Core.Label
+/// could be extended with this SMTPat to proof the
+/// forward secrecy security properties.
+val principal_state_corrupt_implies_principal_corrupt:
+  tr:trace -> prin:principal -> si:state_id -> 
+  Lemma
+  (requires
+    trace_invariant tr /\
+    is_corrupt tr (principal_state_label prin si)
+  )
+  (ensures
+    is_corrupt tr (principal_label prin)
+  )
+  [SMTPat (trace_invariant tr); SMTPat (is_corrupt tr (principal_state_label prin si))]
+let principal_state_corrupt_implies_principal_corrupt tr prin si = 
+  reveal_opaque (`%principal_state_label) (principal_state_label);
+  reveal_opaque (`%principal_label) (principal_label);
+  reveal_opaque (`%pre_is_corrupt) (pre_is_corrupt);
+  normalize_term_spec is_corrupt;
+  
+  // The following code is not needed for the proof.
+  // It just shows what we need to show to prove the lemma.
+  assert(is_corrupt tr (principal_state_label prin si));
+  assert(exists prin' si'. pre_can_flow (S prin si) (S prin' si'));
+  assert(pre_is_corrupt tr (S prin si));
+  assert(exists prin' si'. (S prin si) `pre_can_flow` (S prin' si') /\
+    was_corrupt tr prin' si');
+  assert(exists prin' si'. (P prin) `pre_can_flow` (S prin' si') /\
+    was_corrupt tr prin' si');
+  ()
```

Comment:
This should be deducible from
```fstar

val principal_flow_to_principal_state:
  tr:trace -> prin:principal -> sess_id:state_id ->
  Lemma
  (ensures (principal_label prin) `can_flow tr` (principal_state_label prin sess_id))
  [SMTPat ((principal_label prin) `can_flow tr` (principal_state_label prin sess_id))]
```
and
```fstar

val flow_to_public_eq:
  tr:trace -> l:label ->
  Lemma
  (ensures l `can_flow tr` public <==> is_corrupt tr l)
  [SMTPat (l `can_flow tr` public)]
```
So I am guessing this would work (although I didn't test):
```fstar
let principal_state_corrupt_implies_principal_corrupt tr prin si =
  // Triggers principal_flow_to_principal_state
  assert(principal_label prin `can_flow tr` principal_state_label prin si);
  // Triggers flow_to_public_eq
  assert(principal_state_label prin si `can_flow tr` public);
  // Triggers flow_to_public_eq
  assert(principal_label prin `can_flow tr` public)
```

Reply:
Thanks that works! Do you see a way to trigger these SMTPats directly in the security property proof so that we don't need the lemma? Or do you think that would be a useful lemma for the core DY*?

Reply:
Maybe a lemma like this would be useful to add in DY.Core.Label.Derived (didn't test but should work)

```fstar
val can_flow_propagates_is_corrupt:
  tr:trace -> l1:label -> l2:label ->
  Lemma
  (requires
    is_corrupt tr l2 /\
    l1 `can_flow tr` l2
  )
  (ensures is_corrupt l1)
  [SMTPat (is_corrupt tr l2); SMTPat (l1 `can_flow tr` l2)]
let can_flow_propagates_is_corrupt tr l1 l2 =
  flow_to_public_eq tr l1;
  flow_to_public_eq tr l2
```
So that in you proofs, you simply need to do the `can_flow` assert